workorder-job-repeater.lua

```diff
@@ -0,0 +1,278 @@
+--]]
+local function addJobToPostings(job)
+    local addedIndex = false
+    -- Find the first free empty postings to add a job to
+    for index, posting in ipairs(df.global.world.jobs.postings) do
+        if posting.job == nil then
+            -- It's free real estate!
+            posting.job = job
+            posting.anon_1 = 0
+            posting.flags.dead = false
+
+            job.posting_index = posting.idx
+            addedIndex = posting.idx
+            break
+        end
+    end
+    if addedIndex then
+        log(DEBUG, "Added job " .. job.id .. " to postings as index " .. addedIndex)
+    else
+        log(DEBUG, "Couldn't find posting slot to add in job " .. job.id)
+    end
+
+    return addedIndex
+end
+
+local function repeatJob(order, job)
+    if not (job and order) then
+        -- should never ever happen #hope
+        qerror("job or order is nil")
+        return
+    end
+
+    if only_infty and 0 ~= order.amount_total then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. "): order.amount_total > 0")
+        return
+    end
+
+    if not (order.status.validated and order.status.active) then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id .. " because of order.status")
+        return
+    end
+
+    -- for debug purposes: pause the game
+    --df.global.pause_state = true
+
+    log(DEBUG, "Repeat job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id)
+
+    -- we need a clone because original(which itself is a copy) is destroyed or something.
+    -- cloneJobStruct doesn't copy flags over.
+    local by_manager = job.flags.by_manager
+    local do_now = job.flags.do_now
+    job = dfhack.job.cloneJobStruct(job)
+    job.flags.by_manager = by_manager
+    job.flags.do_now = do_now
+
+    -- step 0: adjust what we can right now
+    job.flags.working = false
+    job.completion_timer = -1
+
+    -- remove old items we worked upon
+    log(DEBUG, "Freeing old items")
+    for _, item in pairs(job.items) do
+        item:delete()
+    end
+    job.items:resize(0)
+
+    local workshop = nil
+    local ixs = {}
+    for ix, gref in ipairs(job.general_refs) do
+        if gref:getType() == df.general_ref_type.BUILDING_HOLDER then
+            local building = gref:getBuilding()
+            if building then
+                log(DEBUG, "Building from gref: " .. building.id)
+                workshop = building
+            else
+                -- should never ever happen, I think
+                log(WARN, "Building from gref not found: ", gref.building_id)
+            end
+        else
+            ixs[ix] = true
+        end
+    end
+    for ix, gref in pairs(ixs) do
+        log(DEBUG, "Removing gref ", gref)
+        gref:delete()
+        job.general_refs:erase(ix)
+    end
+    log(DEBUG, "#gref: " .. #job.general_refs .. " (should be 1)")
```

Comment:
I don't think this will work as expected. `ixs` will contain the indices to remove, but what those indices point to will change as we call `job.general_refs:erase()`. also, the second loop assumes the value in `ixs` is the general ref, but the first loop actually just sets the value to `true`. My guess is that this code has never actually removed any general references before. Is there a way you can get multiple general refs on a job so you can test this?